open-ce/build_tree.py

```diff
@@ -293,32 +316,53 @@ def __init__(self,
         self._conda_env_files = dict()
         self._test_feedstocks = dict()
+        self._initial_package_indices = []
 
         # Create a dependency tree that includes recipes for every combination
         # of variants.
         self._possible_variants = utils.make_variants(python_versions, build_types, mpi_types, cuda_versions)
         self.build_commands = []
         for variant in self._possible_variants:
             try:
-                build_commands, external_deps, test_feedstocks = self._create_all_commands(variant)
+                build_commands, external_deps = self._create_all_commands(variant)
             except OpenCEError as exc:
                 raise OpenCEError(Error.CREATE_BUILD_TREE, exc.msg) from exc
             variant_string = utils.variant_string(variant["python"], variant["build_type"],
                                                   variant["mpi_type"], variant["cudatoolkit"])
             self._external_dependencies[variant_string] = external_deps
-            self._test_feedstocks[variant_string] = test_feedstocks
-            validate_config.validate_build_tree(build_commands, external_deps)
-
-            installable_packages = get_installable_packages(build_commands, external_deps)
-            self._conda_env_files[variant_string] = CondaEnvFileGenerator(installable_packages)
 
             # Add dependency tree information to the packages list and
             # remove build commands from build_commands that are already in self.build_commands
-            build_commands = _add_build_command_dependencies(build_commands, self.build_commands,
-                                        len(self.build_commands))
+            build_commands, package_indices = _add_build_command_dependencies(build_commands, self.build_commands,
+                                                                              len(self.build_commands))
             self.build_commands += build_commands
+            self._detect_cycle()
+
+            # If the packages argument is provided, find the indices into the build_commands for all
+            # of the packages that were requested.
+            variant_package_indices = []
+            if packages:
+                for package in packages:
+                    if package in package_indices:
+                        variant_package_indices += package_indices[package]
+                    else:
+                        print("INFO: No recipes were found for " + package + " for variant " + variant_string)
+            else:
+                for package in package_indices:
+                    variant_package_indices += package_indices[package]
+            self._initial_package_indices += variant_package_indices
 
-        self._detect_cycle()
+            validate_config.validate_build_tree(self.build_commands, external_deps, variant_package_indices)
+            installable_packages = get_installable_packages(self.build_commands, external_deps, variant_package_indices)
```

Comment:
Should it be just `build_commands` here? `self.build_commands` will have all of the commands for all the variants seen till now. I'm not very sure, but may be a point to re-think.

Reply:
I am purposely using `self.build_commands` here because `variant_package_indices` might have indices that point to build_commands that were generated by previous variants. This is because there might be duplicate packages between variants. When that is the case, we use the original one that was added, and don't add the newer duplicates.

Reply:
Okay.